WalletWasabi.Tests/UnitTests/Wabisabi/Backend/CoordinatorTests.cs

```diff
@@ -32,37 +36,46 @@ public async Task CanCancelAsync()
 			await IoHelpers.TryDeleteDirectoryAsync(workDir);
 			CoordinatorParameters coordinatorParameters = new(workDir);
 
-			using WabiSabiCoordinator coordinator = new(coordinatorParameters, new MockRpcClient());
+			using WabiSabiCoordinator coordinator = new(coordinatorParameters, NewMockRpcClient());
 			using CancellationTokenSource cts = new();
 			cts.Cancel();
 			await coordinator.StartAsync(cts.Token);
 			await coordinator.StopAsync(CancellationToken.None);
 
-			using WabiSabiCoordinator coordinator2 = new(coordinatorParameters, new MockRpcClient());
+			using WabiSabiCoordinator coordinator2 = new(coordinatorParameters, NewMockRpcClient());
 			using CancellationTokenSource cts2 = new();
 			await coordinator2.StartAsync(cts2.Token);
 			cts2.Cancel();
 			await coordinator2.StopAsync(CancellationToken.None);
 
-			using WabiSabiCoordinator coordinator3 = new(coordinatorParameters, new MockRpcClient());
+			using WabiSabiCoordinator coordinator3 = new(coordinatorParameters, NewMockRpcClient());
 			using CancellationTokenSource cts3 = new();
 			var t = coordinator3.StartAsync(cts3.Token);
 			cts3.Cancel();
 			await t;
 			await coordinator3.StopAsync(CancellationToken.None);
 
-			using WabiSabiCoordinator coordinator4 = new(coordinatorParameters, new MockRpcClient());
+			using WabiSabiCoordinator coordinator4 = new(coordinatorParameters, NewMockRpcClient());
 			await coordinator4.StartAsync(CancellationToken.None);
 			using CancellationTokenSource cts4 = new();
 			cts4.Cancel();
 			await coordinator4.StopAsync(cts4.Token);
 
-			using WabiSabiCoordinator coordinator5 = new(coordinatorParameters, new MockRpcClient());
+			using WabiSabiCoordinator coordinator5 = new(coordinatorParameters, NewMockRpcClient());
 			await coordinator5.StartAsync(CancellationToken.None);
 			using CancellationTokenSource cts5 = new();
 			t = coordinator5.StopAsync(cts5.Token);
 			cts5.Cancel();
 			await t;
 		}
+
+		private static IRPCClient NewMockRpcClient()
+		{
+			var mockRpcClient = new Mock<IRPCClient>();
+			mockRpcClient.Setup(c => c.Network).Returns(Network.Main);
```

Comment:
If you comment out this line or the following one, the test will still pass even though many NREs can be seen in logs. 

That's probably the worst thing that can happen in a test: To persuade somebody that a test passes even though it *might* be broken.

Reply:
@nopara73 @molnard How can this be fine?

Reply:
I think nobody understands what you mean. Would you like to elaborate?

Reply:
Isn't the point is that the functions in the tests uses the custom values of the ```mockRpcClient```  to pass?
The tests passes with a ```new MockRpcClient()```, which has
```	public Network Network { get; set; } = Network.RegTest;```
at line 24. I assume it uses the testnet of some sort.
But here this mock's network will give us ```Network.Main```.
This simply means that the tests should pass while runing on Main too, right?

Reply:
There are [two types of mocks](https://github.com/Moq/moq4/wiki/Quickstart#customizing-mock-behavior): Loose or strict in Moq library.

Command `var mockRpcClient = new Mock<IRPCClient>();` uses loose mock (as it is default), to have a strict mock, one need to write `var mockRpcClient = new Mock<IRPCClient>(MockBehavior.Strict);`.

The documentation [explains it](https://github.com/Moq/moq4/wiki/Quickstart#customizing-mock-behavior) very well but the gist is: 

* When you use a *strict mock* then an exception is thrown for anything that doesn't have a corresponding expectation (meaning the `setup` method - e.g. `mockRpcClient.Setup(c => c.Network).Returns(Network.Main);`).
* When you use a *loose mock* an exception is never thrown and default values or empty arrays, enumerables, etc. are returned for accessed methods/fields/properties if no expectation is set for the method/field/property.

What does it mean in practice? Consider this:

```csharp
mockRpcClient.Setup(c => c.Network).Returns(Network.Main);
```

When using a loose mock, you don't know whether that `Network` property was accessed or not in your test. You just know that IF accessed `Network.Main` will be returned. IF NOT acccessed, `null` will be returned instead (and the implicit hope is that will fail but ... it may not). If a strict mock were used here and you did not use `Network` property in the test code you would get an exception and your test would fail. If it happens that you forget to add `mockRpcClient.Setup(c => c.Network).Returns(Network.Main);` you would get an exception too.

I have [pointed this out](https://github.com/zkSNACKs/WalletWasabi/pull/5290#discussion_r586489427) in the original PR which added Moq library but Lucas disagrees. I disagree with Lucas as my fear is that tests will dissynchronize with the production code and nobody will know which lines of test code are actually needed and which are not and thus it will require much more debugging to understand why a test fails if something changes.

The greatest danger I can see is that a loose mock can return a default value in a test instead of some pre-defined value and you will persuade yourself that a refactoring was done well but it will test *something else*.

Hopefully, I made myself clear.

Reply:
> Isn't the point is that the functions in the tests uses the custom values of the `mockRpcClient` to pass?

I'm not sure I understand fully but yes, a mock is an object with predefined behavior which is used instead of a real object to make a test shorter and you have greater control over test behavior. Also, very beneficial for edge cases too (which are typically hard to simulate in real testing or in integration testing).

> The tests passes with a `new MockRpcClient()`, which has
> ` public Network Network { get; set; } = Network.RegTest;`
> at line 24. I assume it uses the testnet of some sort.
> But here this mock's network will give us `Network.Main`.
> This simply means that the tests should pass while runing on Main too, right?

Yes, but my point is that you expect `Network.Main` to be returned somewhere but it may not be the case as the `mock.Network` may never be actually accessed in the test itself because somebody changed implementation. So you see a line in code and you basically do not know if it is useful or not.

Reply:
```suggestion
			var mockRpcClient = new Mock<IRPCClient>(MockBehavior.Strict);
			mockRpcClient.Setup(c => c.Network).Returns(Network.Main);
			return mockRpcClient.Object;
```
This way both tests passes. 
Tested without the network setup too, got an error as expected, but there was no error because of the missing exception of ```EstimateSmartFeeAsync```.

Reply:
You can run this test in CoordinatorTests.cs:

```csharp
[Fact]
public async Task CoordinatorTestAsync()
{
	Logger.SetModes(LogMode.Debug, LogMode.File);
	Logger.TurnOn();
	var workDir = Common.GetWorkDir();
	await IoHelpers.TryDeleteDirectoryAsync(workDir);
	CoordinatorParameters coordinatorParameters = new(workDir);

	var mockRpcClient = new Mock<IRPCClient>(MockBehavior.Strict);
	mockRpcClient.Setup(c => c.Network).Returns(Network.Main);
	// mockRpcClient.Setup(c => c.EstimateSmartFeeAsync(It.IsAny<int>(), It.IsAny<EstimateSmartFeeMode>()))
	//	.ReturnsAsync(new EstimateSmartFeeResponse { Blocks = 5, FeeRate = new FeeRate(100m) });

	using WabiSabiCoordinator coordinator2 = new(coordinatorParameters, mockRpcClient.Object);
	using CancellationTokenSource cts2 = new();
	await coordinator2.StartAsync(cts2.Token);
	await Task.Delay(5000);
	cts2.Cancel();
	await coordinator2.StopAsync(CancellationToken.None);
}
```

and you will see an exception in debug window or in your test logs. It does not fail the test because `Arena` is run on a backround thread.

Reply:
Sorry, but im kinda, we were looking at the code, trying out some new lines, but we are staring to the void, so at this point:
What is actually our goal here, if there is one?
To achive that the test fails in case of a thrown exception?

Reply:
In PeriodicRunner.cs we are catching every exception we cannot handle and logging it out. Therefore the test says it's all good because there is no Assert in it.
`			await Assert.ThrowsAnyAsync<MockException>(() => coordinator2.StartAsync(cts2.Token));
`
with this and not catching every exception we can detect if the mocked (or real) RpcClient could run corretly the EstimateSmartFeeAsync method.
But i'm pretty sure we should not fiddle with the PeriodicRunner.cs, so do we have any alternative solution to this?

Reply:
> What is actually our goal here, if there is one?

You got to the specifics here. So if you actually want to achieve that the test fails if `EstimateSmartFeeAsync` is not called, you should consult https://github.com/Moq/moq4/wiki/Quickstart#verification to verify that `EstimateSmartFeeAsync` is called. This is specific to this particular test case.

Anyway, before we got deep into the topic, the goal was to say that strict mocks allows you to know that a test failed sooner. Whereas loose mocks allows you to setup your mock in a way that does not represent reality which I find really bad + it also allows to sneak in subtle bugs (this is a deal breaker for me).

Reply:
Thanks @kiminuo for the elaborate explanation, it makes sense. This PR's decision follows the way how I used the mock classes, I explicitly mocked what I needed and mocked to avoid noisy logs and I didn't rely on default mocking behavior.